### ISSUE TITLE
Stopping and then resuming the downloading procedure may cause the ultimate completed file corrupt.

### ISSUE THREAD

**Issue by hongyi-zhao** [pyIDM/PyIDM#99]
Hi,

I try to test [this large file](http://mirrors.ctan.org/systems/texlive/Images/texlive2020-20200406.iso) downloading with pyidm. After the downloading completed successfully, I compute the md5sum of the downloaded file as following:

```
werner@X10DAi-01:~/Downloads$ md5sum texlive2020-20200406.iso
be137c6851731fb336e8f0df07de5ef1  texlive2020-20200406.iso
```

The above result is different from the [given value on the official website](http://mirrors.ctan.org/systems/texlive/Images/texlive2020-20200406.iso.md5) which is as follows:

`a31c500c7830d901363c5fa0e65cc18d  texlive2020-20200406.iso`

I've tried several times for doing the same test with pyidm. And I found that if I click on the stop and then resume during the downloading process for several times, the ultimate downloaded file is very likely error and has a mismatch md5sum with the official supplied one.

To do the further test on the above problem, I re-check with the same method using [this relative small file](https://www.python.org/ftp/python/3.8.3/Python-3.8.3.tar.xz). And find that some times the result are wrong. See the following for more info:

```
werner@X10DAi-01:~/Downloads$ md5sum Python-3.8.3.tar.xz 
3000cf50aaa413052aef82fd2122ca78  Python-3.8.3.tar.xz
werner@X10DAi-01:~/Downloads$ md5sum Python-3.8.3.tar.xz 
3000cf50aaa413052aef82fd2122ca78  Python-3.8.3.tar.xz
werner@X10DAi-01:~/Downloads$ md5sum Python-3.8.3.tar.xz 
477234541fd6d0d9ced828d25d193ac4  Python-3.8.3.tar.xz
werner@X10DAi-01:~/Downloads$ md5sum Python-3.8.3.tar.xz 
3000cf50aaa413052aef82fd2122ca78  Python-3.8.3.tar.xz
werner@X10DAi-01:~/Downloads$ md5sum Python-3.8.3.tar.xz 
3000cf50aaa413052aef82fd2122ca78  Python-3.8.3.tar.xz

```
As you can see, I tried 5 times, only 4 of them I obtained a correct one.

To summarize:

Stopping and then resuming the downloading procedure may cause the ultimate completed file corrupt. The larger the file is, the higher probability it will be corrupted.

If not intentionally stop and then resume the downloading procedure, the ultimate downloaded file has the highest intact/correct probability.

So, it seems the current segmentation and block combination algorithm used by pyidm is not so robust.  A side note, I also tried the same tests with [flareget](https://flareget.com/) and [IDM](https://www.internetdownloadmanager.com/).  The results show that only IDM can pass the test and obtain an ultimate correct downloaded file regardless the file is huge or tiny.

Regards

**Comment by pyIDM** [comment]
just tested the small file link and i can confirm there is a bug when you stop / resume download and everything goes well if download the file one shot without interruption
i am still looking for the cause for this behavior

**Comment by hongyi-zhao** [comment]
I think you can do a byte-by-byte comparison between the corrupt and intact one to find some clue. Presumably the problem is caused by some incorrect padding among the segments/blocks.

**Comment by hongyi-zhao** [comment]
Further test report on this problem:

[Uget ](https://ugetdm.com/features/) doesn't have this problem.

**Comment by pyIDM** [comment]
I tested PyIDM with one connection with stop/resume and have no problem, only getting this problem when multi-connections or more than one worker is running
still conducting more tests

**Comment by hongyi-zhao** [comment]
> I tested PyIDM with one connection with stop/resume and have no problem, only getting this problem when multi-connections or more than one worker is running

Obviously, this means that there must be some bug in the multiple segment combination algorithm used by PyIDM.

> still conducting more tests

**Comment by pyIDM** [comment]
Just added a fix, for this issue,  hope it works well for you

**Comment by pyIDM** [comment]
@hongyi-zhao 
hold on, it needs another fix in thread manager

**Comment by pyIDM** [comment]
Just tested again a couple of times and this fix should get it right this time.

**Comment by hongyi-zhao** [comment]
> Just added a fix, for this issue, hope it works well for you

I hope it works for __all__, :-)

**Comment by pyIDM** [comment]
hope so :)

**Comment by hongyi-zhao** [comment]
I confirm the recent commits solve the problem.

**Comment by pyIDM** [comment]
> I confirm the recent commits solve the problem reported here.

Great i will close this issue
